File: crates/pty-testkit/src/session.rs
//! A Playwright-style terminal test session, backed by **libghostty** as the
//! terminal emulator and a real PTY (`portable-pty`) as the process host.
//!
//! This is the Rust port of the pty project's `src/testing/session.ts`
//! (spawn mode). Instead of `@xterm/headless`, the PTY byte stream is fed into
//! a libghostty `Terminal`, and "screenshots" are produced by libghostty's
//! formatter.
//!
//! ```no_run
//! use pty_testkit::Session;
//! let mut s = Session::spawn("bash", &["--norc", "--noprofile"], Default::default()).unwrap();
//! s.wait_for_text("$", 5000).unwrap();
//! s.type_str("echo hello\r");
//! s.wait_for_text("hello", 5000).unwrap();
//! s.close();
//! ```

use std::collections::BTreeMap;
use std::io::{Read, Write};
use std::path::PathBuf;
use std::sync::mpsc::{self, Receiver, TryRecvError};
use std::time::{Duration, Instant};

use portable_pty::{native_pty_system, CommandBuilder, MasterPty, PtySize};

use pty_core::keys::{resolve_key, KeyError};
use pty_terminal::{Screenshot, TerminalActor};

/// Options for [`Session::spawn`]. Mirrors the TS `SpawnOptions`.
#[derive(Debug, Clone, Default)]
pub struct SpawnOptions {
    /// Terminal height in rows. Default: 24.
    pub rows: Option<u16>,
    /// Terminal width in columns. Default: 80.
    pub cols: Option<u16>,
    /// Working directory for the spawned process.
    pub cwd: Option<PathBuf>,
    /// Extra environment variables, merged over the inherited environment.
    pub env: Vec<(String, String)>,
}

/// Build the environment for a spawned `pty` process: the caller's base env
/// merged with `opts_env`, minus the harness's own pty-internal context.
///
/// Port of `buildSpawnEnv` from `session.ts`. Always scrubs
/// `PTY_SERVER_CONFIG` and `PTY_SESSION`; scrubs `PTY_ROOT` /
/// `PTY_SESSION_DIR` only when the caller didn't set them explicitly.
pub fn build_spawn_env(
    base: &[(String, String)],
    opts_env: &[(String, String)],
) -> BTreeMap<String, String> {
    let mut env: BTreeMap<String, String> = base.iter().cloned().collect();

    let opts_has = |k: &str| opts_env.iter().any(|(ek, _)| ek == k);
    for (k, v) in opts_env {
        env.insert(k.clone(), v.clone());
    }
    env.remove("PTY_SERVER_CONFIG");
    env.remove("PTY_SESSION");
    if !opts_has("PTY_ROOT") {
        env.remove("PTY_ROOT");
    }
    if !opts_has("PTY_SESSION_DIR") {
        env.remove("PTY_SESSION_DIR");
    }
    env
}

/// How long the `_default` waits allow, matching the Node package.
pub const DEFAULT_WAIT_MS: u64 = 10_000;

/// A spawned terminal session driving a real process through a PTY, with a
/// libghostty terminal tracking the screen state.
pub struct Session {
    /// The libghostty terminal, owned by the actor (query answers, mode
    /// tracking, and the one serializer shared with the daemon).
    actor: TerminalActor,
    writer: Box<dyn Write + Send>,
    rx: Receiver<Vec<u8>>,
    backing: Backing,
    rows: u16,
    cols: u16,
}

/// What is on the other end: a process this library started, or a session a
/// `pty` daemon owns.
enum Backing {
    Pty {
        master: Box<dyn MasterPty + Send>,
        child: Box<dyn portable_pty::Child + Send + Sync>,
    },
    Server(Box<crate::server::ServerBacking>),
}

/// How [`Session::server`] creates or finds a session.
#[derive(Debug, Clone, Default)]
pub struct ServerOptions {
    /// The session id. A random one when absent.
    pub name: Option<String>,
    pub rows: Option<u16>,
    pub cols: Option<u16>,
    pub cwd: Option<String>,
    /// Passed through as `--env KEY=VALUE`.
    pub env: Vec<(String, String)>,
    /// The registry to use. A temporary one when absent, removed on close.
    pub root: Option<std::path::PathBuf>,
}

impl Session {
    /// Spawn a process in a direct PTY. Use for CLI tools, TUI apps, or any
    /// process where you send input and check screen output.
    pub fn spawn(command: &str, args: &[&str], opts: SpawnOptions) -> std::io::Result<Session> {
        let rows = opts.rows.unwrap_or(24);
        let cols = opts.cols.unwrap_or(80);

        // The terminal actor answers queries (DA1, DSR, …) into its reply
        // buffer; `pump` flushes them back to the PTY, so programs like fish
        // that block on a DA1 reply start promptly.
        let actor = TerminalActor::new(rows, cols, pty_terminal::actor::DEFAULT_SCROLLBACK);

        // Real PTY + child process.
        let pty_system = native_pty_system();
        let pair = pty_system
            .openpty(PtySize {
                rows,
                cols,
                pixel_width: 0,
                pixel_height: 0,
            })
            .map_err(std::io::Error::other)?;

        let mut cmd = CommandBuilder::new(command);
        cmd.args(args);
        if let Some(cwd) = &opts.cwd {
            cmd.cwd(cwd);
        }
        // Apply the scrubbed, merged environment explicitly.
        let base: Vec<(String, String)> = std::env::vars().collect();
        let mut env = build_spawn_env(&base, &opts.env);
        // Default TERM to match the TS harness (node-pty used name:
        // "xterm-256color"); a caller-provided TERM in opts.env already won.
        env.entry("TERM".to_string())
            .or_insert_with(|| "xterm-256color".to_string());
        cmd.env_clear();
        for (k, v) in env {
            cmd.env(k, v);
        }

        let child = pair
            .slave
            .spawn_command(cmd)
            .map_err(std::io::Error::other)?;
        // Slave no longer needed in the parent; dropping it avoids a lingering
        // open fd that would keep the master from ever seeing EOF.
        drop(pair.slave);

        let reader = pair
            .master
            .try_clone_reader()
            .map_err(std::io::Error::other)?;
        let writer = pair.master.take_writer().map_err(std::io::Error::other)?;

        // Reader thread: read PTY bytes → channel. libghostty's Terminal is
        // !Send, so it stays on this thread; the thread only ferries bytes.
        let (tx, rx) = mpsc::channel::<Vec<u8>>();
        std::thread::spawn(move || {
            let mut reader = reader;
            let mut buf = [0u8; 8192];
            loop {
                match reader.read(&mut buf) {
                    Ok(0) => break,
                    Ok(n) => {
                        if tx.send(buf[..n].to_vec()).is_err() {
                            break;
                        }
                    }
                    Err(_) => break,
                }
            }
        });

        Ok(Session {
            actor,
            writer,
            rx,
            backing: Backing::Pty {
                master: pair.master,
                child,
            },
            rows,
            cols,
        })
    }

    /// Create a session through the `pty` binary and drive it over its
    /// socket. The screen arrives as frames rather than from a pty this
    /// process owns, so this is what a real client sees.
    ///
    /// The binary is `PTY_BIN`, else `pty` on PATH.
    pub fn server(command: &str, args: &[&str], opts: ServerOptions) -> std::io::Result<Session> {
        let rows = opts.rows.unwrap_or(24);
        let cols = opts.cols.unwrap_or(80);
        let name = opts.name.clone().unwrap_or_else(crate::server::random_id);
        let bin = crate::server::pty_bin();
        let owns_root = opts.root.is_none();
        let root = match opts.root.clone() {
            Some(root) => root,
            None => {
                // Short: a session socket path has to fit 104 bytes.
                let dir = std::env::temp_dir().join(format!("pt-{}", crate::server::random_id()));
                std::fs::create_dir_all(&dir)?;
                dir
            }
        };
        crate::server::spawn_daemon(
            &bin,
            &root,
            &name,
            command,
            args,
            rows,
            cols,
            opts.cwd.as_deref(),
            &opts.env,
        )?;
        Session::connect_at(&bin, root, name, rows, cols, true, owns_root)
    }

    /// Attach to a session that already exists, without owning it: closing
    /// this handle leaves the session running.
    pub fn connect(name: &str, rows: u16, cols: u16, root: std::path::PathBuf) -> std::io::Result<Session> {
        Session::connect_at(
            &crate::server::pty_bin(),
            root,
            name.to_string(),
            rows,
            cols,
            false,
            false,
        )
    }

    /// A second client on the same session, at its own size. The daemon
    /// gives every client the smallest requested geometry.
    pub fn connect_to_existing(other: &Session, rows: u16, cols: u16) -> std::io::Result<Session> {
        let Backing::Server(server) = &other.backing else {
            return Err(std::io::Error::other(
                "connect_to_existing needs a server-mode session",
            ));
        };
        Session::connect_at(
            &server.bin,
            server.root.clone(),
            server.name.clone(),
            rows,
            cols,
            false,
            false,
        )
    }

    fn connect_at(
        bin: &str,
        root: std::path::PathBuf,
        name: String,
        rows: u16,
        cols: u16,
        owned: bool,
        owns_root: bool,
    ) -> std::io::Result<Session> {
        let actor = TerminalActor::new(rows, cols, pty_terminal::actor::DEFAULT_SCROLLBACK);
        let (tx, rx) = mpsc::channel::<Vec<u8>>();
        let (socket, state) = crate::server::connect(&root, &name, rows, cols, tx)?;
        let writer = Box::new(crate::server::DataFramer(socket.try_clone()?));
        Ok(Session {
            actor,
            writer,
            rx,
            backing: Backing::Server(Box::new(crate::server::ServerBacking {
                name,
                root,
                bin: bin.to_string(),
                socket,
                state,
                owned,
                owns_root,
            })),
            rows,
            cols,
        })
    }

    /// The session's id. Empty for a spawned process, which has none.
    pub fn name(&self) -> &str {
        match &self.backing {
            Backing::Server(server) => &server.name,
            Backing::Pty { .. } => "",
        }
    }

    /// The registry this session lives in, for a server-mode session.
    pub fn root(&self) -> Option<&std::path::Path> {
        match &self.backing {
            Backing::Server(server) => Some(&server.root),
            Backing::Pty { .. } => None,
        }
    }

    /// The exit status, once the session has ended.
    pub fn exit_code(&self) -> Option<i32> {
        match &self.backing {
            Backing::Server(server) => server
                .state
                .exited
                .load(std::sync::atomic::Ordering::Acquire)
                .then(|| server.state.exit_code.load(std::sync::atomic::Ordering::Relaxed)),
            Backing::Pty { .. } => None,
        }
    }

    /// Drop this client and open a new one, the way a client that lost its
    /// connection would. The screen is rebuilt from the daemon's replay.
    pub fn reconnect(&mut self) -> std::io::Result<()> {
        let Backing::Server(server) = &mut self.backing else {
            return Err(std::io::Error::other("reconnect needs a server-mode session"));
        };
        let (bin, root, name) = (server.bin.clone(), server.root.clone(), server.name.clone());
        let (owned, owns_root) = (server.owned, server.owns_root);
        let _ = server.socket.shutdown(std::net::Shutdown::Both);
        std::thread::sleep(std::time::Duration::from_millis(100));

        let actor = TerminalActor::new(self.rows, self.cols, pty_terminal::actor::DEFAULT_SCROLLBACK);
        let (tx, rx) = mpsc::channel::<Vec<u8>>();
        let (socket, state) = crate::server::connect(&root, &name, self.rows, self.cols, tx)?;
        self.writer = Box::new(crate::server::DataFramer(socket.try_clone()?));
        self.actor = actor;
        self.rx = rx;
        self.backing = Backing::Server(Box::new(crate::server::ServerBacking {
            name,
            root,
            bin,
            socket,
            state,
            owned,
            owns_root,
        }));
        Ok(())
    }

    /// Drain all currently-available PTY output into the terminal, then flush
    /// any query responses libghostty produced back to the PTY.
    fn pump(&mut self) {
        loop {
            match self.rx.try_recv() {
                Ok(chunk) => {
                    self.actor.write(&chunk);
                }
                Err(TryRecvError::Empty) => break,
                Err(TryRecvError::Disconnected) => break,
            }
        }
        let out = self.actor.take_pty_replies();
        if !out.is_empty() {
            let _ = self.writer.write_all(&out);
            let _ = self.writer.flush();
        }
    }

    // ── Properties ──

    /// Current terminal height in rows. In server mode this is the size the
    /// daemon settled on, which may be smaller than the one requested.
    pub fn rows(&self) -> u16 {
        match &self.backing {
            Backing::Server(server) => server.state.rows.load(std::sync::atomic::Ordering::Relaxed),
            Backing::Pty { .. } => self.rows,
        }
    }

    /// Current terminal width in columns. See [`Session::rows`].
    pub fn cols(&self) -> u16 {
        match &self.backing {
            Backing::Server(server) => server.state.cols.load(std::sync::atomic::Ordering::Relaxed),
            Backing::Pty { .. } => self.cols,
        }
    }

    // ── Input ──

    /// Send raw keystrokes to the process. Use for literal text or escape
    /// sequences.
    pub fn send_keys(&mut self, keys: &str) {
        let _ = self.writer.write_all(keys.as_bytes());
        let _ = self.writer.flush();
    }

    /// Send a named key. Supports modifiers: `"ctrl+c"`, `"alt+x"`, `"shift+a"`.
    pub fn press(&mut self, key_name: &str) -> Result<(), KeyError> {
        let bytes = resolve_key(key_name)?;
        self.send_keys(&bytes);
        Ok(())
    }

    /// Send text to the process. Alias for [`Session::send_keys`].
    pub fn type_str(&mut self, text: &str) {
        self.send_keys(text);
    }

    // ── Screen ──

    /// Capture the current terminal state (drains pending output first).
    pub fn screenshot(&mut self) -> Screenshot {
        self.pump();
        self.actor.screenshot()
    }

    /// The terminal actor behind this session, for typed reads
    /// (`snapshot`, `modes`, `plain`) after a [`Session::screenshot`] pump.
    pub fn actor(&self) -> &TerminalActor {
        &self.actor
    }

    /// The current terminal window title (set by the program via OSC 0/2).
    /// Drains pending output first.
    pub fn title(&mut self) -> String {
        self.pump();
        self.actor.title()
    }

    // ── Waiting ──

    /// Poll until the terminal contains `text`. Returns the matching screenshot.
    pub fn wait_for_text(&mut self, text: &str, timeout_ms: u64) -> Result<Screenshot, String> {
        self.wait_for(
            |ss| ss.text.contains(text),
            timeout_ms,
            &format!("text {text:?}"),
        )
    }

    /// Poll until the terminal no longer contains `text`.
    pub fn wait_for_absent(&mut self, text: &str, timeout_ms: u64) -> Result<Screenshot, String> {
        self.wait_for(
            |ss| !ss.text.contains(text),
            timeout_ms,
            &format!("absence of {text:?}"),
        )
    }

    /// Poll until `predicate` returns true. `description` is used in the
    /// timeout error message.
    pub fn wait_for<F>(
        &mut self,
        predicate: F,
        timeout_ms: u64,
        description: &str,
    ) -> Result<Screenshot, String>
    where
        F: Fn(&Screenshot) -> bool,
    {
        let start = Instant::now();
        let timeout = Duration::from_millis(timeout_ms);
        loop {
            let ss = self.screenshot();
            if predicate(&ss) {
                return Ok(ss);
            }
            if start.elapsed() >= timeout {
                return Err(format!(
                    "Timed out after {timeout_ms}ms waiting for {description}.\nScreen:\n{}",
                    ss.text
                ));
            }
            std::thread::sleep(Duration::from_millis(20));
        }
    }

    // ── Resize ──

    /// Resize the far end and the terminal emulator together.
    ///
    /// In server mode the daemon decides the size — every client asks and
    /// the smallest wins — so `rows()` and `cols()` follow its answer rather
    /// than what was requested here.
    pub fn resize(&mut self, rows: u16, cols: u16) {
        self.rows = rows;
        self.cols = cols;
        match &mut self.backing {
            Backing::Pty { master, .. } => {
                let _ = master.resize(PtySize {
                    rows,
                    cols,
                    pixel_width: 0,
                    pixel_height: 0,
                });
            }
            Backing::Server(server) => server.resize(rows, cols),
        }
        self.actor.resize(cols, rows);
    }

    // ── Lifecycle ──

    /// True once the process on the far end has exited.
    pub fn has_exited(&mut self) -> bool {
        match &mut self.backing {
            Backing::Pty { child, .. } => matches!(child.try_wait(), Ok(Some(_))),
            Backing::Server(server) => server.state.exited.load(std::sync::atomic::Ordering::Acquire),
        }
    }

    /// Stop the session and clean up after it.
    ///
    /// A spawned process is killed. A session this handle created is killed
    /// and removed through the `pty` binary, so the daemon records its exit
    /// the way it would for any caller; a session this handle merely
    /// connected to is left running.
    pub fn close(&mut self) {
        match &mut self.backing {
            Backing::Pty { child, .. } => {
                let _ = child.kill();
                let _ = child.wait();
            }
            Backing::Server(server) => server.close(),
        }
    }
}

impl Drop for Session {
    fn drop(&mut self) {
        match &mut self.backing {
            Backing::Pty { child, .. } => {
                let _ = child.kill();
            }
            Backing::Server(server) => server.close(),
        }
    }
}

File: packages/pty-testing/src/session.ts
/**
 * A terminal session you can type at and take pictures of.
 *
 * The session itself is hosted by the `pty` binary; this talks to its daemon
 * over the session socket. So what a test sees here is what any client sees.
 */

import { execFile, execFileSync } from "node:child_process";
import { existsSync, mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { connect, type Socket } from "node:net";

import { resolveKey } from "./keys.js";
import {
  MessageType,
  PacketReader,
  decodeExit,
  decodeGeometry,
  encodeAttach,
  encodeData,
  encodePeek,
  encodeResize,
} from "./protocol.js";

/** The environment variables a session must not inherit from the test run. */
const SCRUBBED = [
  "PTY_SESSION",
  "PTY_SESSION_GENERATION",
  "PTY_SESSION_DIR",
  "PTY_REAP_ON_EXIT",
  "NO_COLOR",
];

export interface SpawnOptions {
  rows?: number;
  cols?: number;
  cwd?: string;
  env?: Record<string, string>;
  /** The session id. A random one when absent. */
  name?: string;
}

export interface Screenshot {
  /** The visible lines, with trailing blank ones removed. */
  lines: string[];
  /** Those lines joined with newlines. */
  text: string;
  /** The same screen with its escape sequences intact. */
  ansi: string;
}

/** How long the waits allow when no timeout is given. */
export const DEFAULT_TIMEOUT_MS = 10_000;
const POLL_MS = 50;

let engineChecked = false;

/**
 * The `pty` to drive: `PTY_BIN`, else `pty` on PATH.
 *
 * The first call refuses a binary that is not the Rust one, because this
 * package is written against its behaviour. `PTY_TESTING_ALLOW_NODE=1` says
 * you meant it.
 */
export function ptyBin(): string {
  const bin = process.env.PTY_BIN || "pty";
  if (!engineChecked) {
    engineChecked = true;
    if (process.env.PTY_TESTING_ALLOW_NODE !== "1") {
      let version = "";
      try {
        version = execFileSync(bin, ["version"], { encoding: "utf8" }).trim();
      } catch (e) {
        throw new Error(`could not run "${bin} version": ${(e as Error).message}`);
      }
      if (!version.includes("-rust")) {
        throw new Error(
          `${bin} reports ${version}, which is not the Rust pty. ` +
            `Set PTY_BIN, or PTY_TESTING_ALLOW_NODE=1 if you meant this one.`,
        );
      }
    }
  }
  return bin;
}

function randomId(): string {
  const alphabet = "23456789abcdefghjkmnpqrstuvwxyz";
  let out = "";
  for (let i = 0; i < 8; i++) out += alphabet[Math.floor(Math.random() * alphabet.length)];
  return out;
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

export class Session {
  #socket: Socket | null = null;
  #ansi = "";
  #rows: number;
  #cols: number;
  #exitCode: number | null = null;
  #exited = false;

  private constructor(
    readonly name: string,
    readonly root: string,
    private readonly bin: string,
    private readonly ownsSession: boolean,
    private readonly ownsRoot: boolean,
    rows: number,
    cols: number,
  ) {
    this.#rows = rows;
    this.#cols = cols;
  }

  get rows(): number {
    return this.#rows;
  }
  get cols(): number {
    return this.#cols;
  }
  get hasExited(): boolean {
    return this.#exited;
  }
  get exitCode(): number | null {
    return this.#exitCode;
  }

  /** Create a session and attach to it. */
  static async spawn(command: string, args: string[] = [], opts: SpawnOptions = {}): Promise<Session> {
    const bin = ptyBin();
    const rows = opts.rows ?? 24;
    const cols = opts.cols ?? 80;
    const name = opts.name ?? randomId();
    // Short, because a session socket path has to fit 104 bytes.
    const root = mkdtempSync(join(tmpdir(), "pt-"));

    const argv = ["run", "-d", "-e", "--no-display-name", "--id", name];
    if (opts.cwd) argv.push("--cwd", opts.cwd);
    argv.push("--rows", String(rows), "--cols", String(cols));
    for (const [k, v] of Object.entries(opts.env ?? {})) argv.push("--env", `${k}=${v}`);
    argv.push("--", command, ...args);

    const env: NodeJS.ProcessEnv = { ...process.env, PTY_ROOT: root };
    for (const key of SCRUBBED) delete env[key];
    await new Promise<void>((resolve, reject) => {
      execFile(bin, argv, { env }, (err, _out, stderr) => {
        if (err) reject(new Error(`pty run failed: ${String(stderr).trim() || err.message}`));
        else resolve();
      });
    });

    const session = new Session(name, root, bin, true, true, rows, cols);
    await session.attach();
    return session;
  }

  /** Attach to a session that already exists, without owning it. */
  static async connect(
    name: string,
    opts: { rows?: number; cols?: number; root: string },
  ): Promise<Session> {
    const session = new Session(
      name,
      opts.root,
      ptyBin(),
      false,
      false,
      opts.rows ?? 24,
      opts.cols ?? 80,
    );
    await session.attach();
    return session;
  }

  /** A second client on the same session, at its own size. */
  static async connectToExisting(
    other: Session,
    opts: { rows?: number; cols?: number } = {},
  ): Promise<Session> {
    return Session.connect(other.name, { ...opts, root: other.root });
  }

  /** Open the socket and join the session. Resolves on the first screen. */
  async attach(): Promise<void> {
    const path = join(this.root, `${this.name}.sock`);
    const deadline = Date.now() + 15_000;
    while (!existsSync(path)) {
      if (Date.now() > deadline) throw new Error(`session "${this.name}" never came up`);
      await sleep(20);
    }

    const socket = connect(path);
    this.#socket = socket;
    const reader = new PacketReader();
    let sawScreen = false;
    let onScreen: (() => void) | null = null;
    const screen = new Promise<void>((resolve) => {
      onScreen = resolve;
    });

    socket.on("data", (chunk: Buffer) => {
      let packets;
      try {
        packets = reader.feed(chunk);
      } catch {
        socket.destroy();
        return;
      }
      for (const packet of packets) {
        switch (packet.type) {
          case MessageType.Screen:
          case MessageType.Data:
            this.#ansi += packet.payload.toString("utf8");
            if (packet.type === MessageType.Screen && !sawScreen) {
              sawScreen = true;
              onScreen?.();
            }
            break;
          case MessageType.Geometry: {
            const { rows, cols } = decodeGeometry(packet.payload);
            this.#rows = rows;
            this.#cols = cols;
            break;
          }
          case MessageType.Exit:
            this.#exitCode = decodeExit(packet.payload);
            this.#exited = true;
            break;
        }
      }
    });
    socket.on("close", () => {
      onScreen?.();
    });
    socket.on("error", () => {
      onScreen?.();
    });

    await new Promise<void>((resolve, reject) => {
      socket.once("connect", () => resolve());
      socket.once("error", reject);
    });
    socket.write(encodeAttach(this.#rows, this.#cols));
    // The daemon replays the screen straight after ATTACH; wait for it so a
    // screenshot taken immediately is not empty.
    await Promise.race([screen, sleep(2000)]);
  }

  /** Drop the connection and open a new one, as a client that lost it would. */
  async reconnect(): Promise<void> {
    this.#socket?.destroy();
    this.#socket = null;
    this.#ansi = "";
    await sleep(100);
    await this.attach();
  }

  /** Send text as typed. */
  sendKeys(text: string): void {
    this.#socket?.write(encodeData(text));
  }

  /** Alias for {@link Session.sendKeys}. */
  type(text: string): void {
    this.sendKeys(text);
  }

  /** Send a named key: `ctrl+c`, `C-u`, `return`, `up`. */
  press(key: string): void {
    this.sendKeys(resolveKey(key));
  }

  /** Ask for a new size. The daemon gives every client the smallest asked for. */
  resize(rows: number, cols: number): void {
    this.#rows = rows;
    this.#cols = cols;
    this.#socket?.write(encodeResize(rows, cols));
  }

  /** The screen as it is now. */
  async screenshot(): Promise<Screenshot> {
    const plain = await this.#peek(true);
    const lines = plain.split("\n");
    while (lines.length > 0 && lines[lines.length - 1].trim() === "") lines.pop();
    return { lines, text: lines.join("\n"), ansi: this.#ansi };
  }

  /** Wait for `text` to appear, and return the screen it appeared on. */
  async waitForText(text: string, timeoutMs = DEFAULT_TIMEOUT_MS): Promise<Screenshot> {
    return this.waitFor((s) => s.text.includes(text), timeoutMs, `text "${text}" to appear`);
  }

  /** Wait for `text` to go away. */
  async waitForAbsent(text: string, timeoutMs = DEFAULT_TIMEOUT_MS): Promise<Screenshot> {
    return this.waitFor((s) => !s.text.includes(text), timeoutMs, `text "${text}" to go away`);
  }

  /** Wait until `predicate` is happy with the screen. */
  async waitFor(
    predicate: (screen: Screenshot) => boolean,
    timeoutMs = DEFAULT_TIMEOUT_MS,
    what = "the condition",
  ): Promise<Screenshot> {
    const deadline = Date.now() + timeoutMs;
    let last = await this.screenshot();
    for (;;) {
      if (predicate(last)) return last;
      if (Date.now() > deadline) break;
      await sleep(POLL_MS);
      last = await this.screenshot();
    }
    throw new Error(`Timed out after ${timeoutMs}ms waiting for ${what}. Screen was:\n${last.text}`);
  }

  /**
   * Stop the session if this handle created it, and clean up after it.
   * A handle that merely connected leaves the session running.
   */
  async close(): Promise<void> {
    this.#socket?.destroy();
    this.#socket = null;
    if (this.ownsSession) {
      const env: NodeJS.ProcessEnv = { ...process.env, PTY_ROOT: this.root };
      for (const verb of ["kill", "rm"]) {
        try {
          execFileSync(this.bin, [verb, this.name], { env, stdio: "ignore" });
        } catch {
          // Already gone is the wanted state.
        }
      }
    }
    if (this.ownsRoot) rmSync(this.root, { recursive: true, force: true });
  }

  /** A short-lived socket that reads the screen without joining. */
  async #peek(plain: boolean): Promise<string> {
    const path = join(this.root, `${this.name}.sock`);
    if (!existsSync(path)) {
      // The session ended; the last screen it saved is the best answer.
      const metaPath = join(this.root, `${this.name}.json`);
      if (existsSync(metaPath)) {
        try {
          const meta = JSON.parse(readFileSync(metaPath, "utf8"));
          if (Array.isArray(meta.lastLines)) return meta.lastLines.join("\n");
        } catch {
          // Fall through to an empty screen.
        }
      }
      return "";
    }
    return new Promise<string>((resolve) => {
      const socket = connect(path);
      const reader = new PacketReader();
      let out = "";
      const done = () => {
        socket.destroy();
        resolve(out);
      };
      socket.on("connect", () => socket.write(encodePeek(plain, false)));
      socket.on("data", (chunk: Buffer) => {
        let packets;
        try {
          packets = reader.feed(chunk);
        } catch {
          done();
          return;
        }
        for (const packet of packets) {
          if (packet.type === MessageType.Screen || packet.type === MessageType.Data) {
            out += packet.payload.toString("utf8");
          }
        }
      });
      socket.on("close", done);
      socket.on("error", done);
      setTimeout(done, 2000);
    });
  }
}
